Silence warnings and deprecations in the console during tests

bmac opened the pull request:
Closes #4416 

#### Before
<img width="555" alt="screen shot 2016-11-17 at 10 04 57 am" src="https://cloud.githubusercontent.com/assets/54056/20394454/8e952678-acad-11e6-9035-adec9651ac70.png">
<img width="552" alt="screen shot 2016-11-17 at 10 05 05 am" src="https://cloud.githubusercontent.com/assets/54056/20394458/8ff56a1e-acad-11e6-8af2-114563f67384.png">


#### After
<img width="554" alt="screen shot 2016-11-17 at 9 57 12 am" src="https://cloud.githubusercontent.com/assets/54056/20394461/92d11d3c-acad-11e6-9a5f-01cc198083bc.png">

stefanpenner commented:
just checking but this doesn't only silence them, but asserts they do or do not happen & happens to silence them right?

pangratz commented on tests/integration/serializers/json-serializer-test.js:
```diff
@@ -987,9 +987,10 @@ test('normalizeResponse ignores unmapped attributes', function(assert) {
-  var post = env.store.serializerFor("post").normalizeResponse(env.store, Post, jsonHash, '1', 'findRecord');
-
-  assert.equal(post.data.attributes.title, "Rails is omakase");
+  assert.expectWarning(function() {
```
Since this tests adds an assertion for an expected warning, you need to use `testInDebug` instead of `test`, since otherwise this test will fail when the ember data tests are run in the production build.

pangratz commented on tests/unit/store/adapter-interop-test.js:
```diff
@@ -836,7 +836,7 @@ test("the promise returned by `_scheduleFetch`, when it rejects, does not depend
 });
 
 test("store._fetchRecord reject records that were not found, even when those requests were coalesced with records that were found", function(assert) {
```
see above: `s/test/testInDebug`

bmac commented:
> just checking but this doesn't only silence them, but asserts they do or do not happen & happens to silence them right?

I added 2 assertions that happens to silence the warnings mainly for the side effect of silencing the warnings. The json-serializer-test.js did not have an existing test for the warning, however the adapter-interop-test [does have an explicit test for the warning](https://github.com/bmac/data/blob/dcea34116972fb349842b7777433186a96a28583/tests/unit/store/adapter-interop-test.js#L878) which my change now duplicates.

For the since almost all of the tests in the errors object trigger a warning I overrode [Ember.warn for that entire module](https://github.com/emberjs/data/pull/4663/files#diff-365a2631b698bbbb4a84586bf51c7d7d).

I also added an ignoreDeprecation in the rest-adapter-test.js which doesn't and any assertions since there [is already a test for that deprecation](https://github.com/bmac/data/blob/dcea34116972fb349842b7777433186a96a28583/tests/integration/adapter/rest-adapter-test.js#L1369-L1383)

stefanpenner commented:
Although a pain, I think the ideal is. error on unexpected warn, error on unexpected deprecation. And require assertions in those places.

Maybe there is a middleground, but hiding warning/deprecations is a slippery slope.

pangratz commented:
We had this once (added in #3834) but somehow this isn't working anymore. At least not the last time I checked (if I remember correctly).

pangratz commented:
I see that there has been some work in #3995.

@bmac do you want to investigate if [this](https://github.com/emberjs/data/blob/7a12c44a5631cbef8ed2f30fe10f1df7b2b6c14b/tests/dummy/config/environment.js#L20-L22) might be working now?

bmac commented:
I have updated this pr to explicitly tests for all of the deprecations and warnings that were ignored in the earlier version of this pr.

@pangratz enabling that flag now causes the build to fail when there is a deprecation warning.

pangratz commented on tests/helpers/setup-ember-dev.js:
```diff
@@ -2,7 +2,7 @@
 import EmberTestHelpers from "ember-dev/test-helper/index";
 
-const AVAILABLE_ASSERTIONS = ['expectAssertion', 'expectDeprecation', 'expectNoDeprecation', 'expectWarning', 'expectNoWarning'];
+const AVAILABLE_ASSERTIONS = ['expectAssertion', 'expectDeprecation', 'expectNoDeprecation', 'expectWarning', 'expectNoWarning', 'ignoreDeprecation'];
```
Seems like `ignoreDeprecation` is not used?

pangratz commented:
Thanks for your work in this @bmac!! 🚀
